java-frontend/src/main/java/org/sonar/java/se/checks/InvariantReturnCheck.java

```diff
@@ -0,0 +1,151 @@
+      returnImmutableType = methodToCheck && (returnType.symbolType().isPrimitive() || returnType.symbolType().is("java.lang.String"));
+    }
+
+    private static boolean isConstructorOrVoid(@Nullable TypeTree returnType) {
+      return returnType == null || returnType.symbolType().isVoid();
+    }
+
+    private static boolean hasAtLeastTwoReturn(MethodTree methodTree) {
+      ReturnCounter visitor = new ReturnCounter();
+      methodTree.accept(visitor);
+      return visitor.returnCount > 1;
+    }
+  }
+
+  private Deque<MethodInvariantContext> methodInvariantContexts = new LinkedList<>();
+
+  @Override
+  public void init(MethodTree methodTree, CFG cfg) {
+    methodInvariantContexts.push(new MethodInvariantContext(methodTree));
+  }
+
+  @Override
+  public void checkEndOfExecutionPath(CheckerContext context, ConstraintManager constraintManager) {
+    MethodInvariantContext methodInvariantContext = methodInvariantContexts.peek();
+    if (!methodInvariantContext.methodToCheck) {
+      return;
+    }
+    SymbolicValue exitValue = context.getState().exitValue();
+
+    if (exitValue != null) {
+      methodInvariantContext.endPaths++;
+      methodInvariantContext.symbolicValues.add(exitValue);
+      PMap<Class<? extends Constraint>, Constraint> constraints = context.getState().getConstraints(exitValue);
+      if (constraints != null) {
+        constraints.forEach(methodInvariantContext.methodConstraints::put);
+      } else {
+        // Relational SV or NOT SV : we can't say anything.
+        methodInvariantContext.avoidRaisingConstraintIssue = true;
+      }
+    }
+  }
+
+  @Override
+  public void checkEndOfExecution(CheckerContext context) {
+    reportIssues(context);
+  }
+
+  @Override
+  public void interruptedExecution(CheckerContext context) {
+    reportIssues(context);
+  }
+
+  private void reportIssues(CheckerContext context) {
+    MethodInvariantContext methodInvariantContext = methodInvariantContexts.pop();
+    if (!methodInvariantContext.methodToCheck) {
+      return;
+    }
+    if (methodInvariantContext.returnImmutableType && methodInvariantContext.symbolicValues.size() == 1 && methodInvariantContext.endPaths > 1) {
+      context.getNode().edges().stream()
+        .findFirst()
+        .map(e -> e.parent().programPoint.syntaxTree())
+        .ifPresent(t -> reportIssue(t, "Refactor this method to not always return the same value.\n", Collections.emptySet()));
+    }
+    if(!methodInvariantContext.avoidRaisingConstraintIssue) {
+      for (Class<? extends Constraint> aClass : methodInvariantContext.methodConstraints.keys()) {
+        Collection<Constraint> constraints = methodInvariantContext.methodConstraints.get(aClass);
+        if(constraints.size() == methodInvariantContext.endPaths && constraints.stream().allMatch(c -> constraints.iterator().next().hasPreciseValue() && constraints.iterator().next().equals(c))) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Split this 197 characters long line (which is greater than 180 authorized). [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)